cloud_functions/climateiq_data_export_cf/main.py

```diff
@@ -211,19 +228,173 @@ def _build_spatialized_model_predictions(
+                geometry.Point(cell["h3_centroid_lon"], cell["h3_centroid_lat"])
+            ),
+            axis=1,
+        )
+    ]
+
+    # Extract rows where the projected H3 cell is not fully contained within the chunk
+    # so we can aggregate prediction values across chunk boundaries.
+    boundary_h3_cells = (
+        spatialized_predictions[
+            spatialized_predictions.apply(
+                lambda row: not row["h3_boundary"].within(chunk_boundary),
+                axis=1,
+            )
+        ]
+        .groupby(["h3_index"])
+        .prediction.agg("mean")
+    )
+
+    # TODO: Add logic for fetching predictions from neighboring chunks for boundary
+    # cells.
+    spatialized_predictions = spatialized_predictions.groupby(
+        ["h3_index"]
+    ).prediction.agg("mean")
+    return spatialized_predictions
+
+
+def _get_neighboring_chunks(
+    study_area_metadata: dict, chunk_metadata: dict, boundary_h3_cells: pd.DataFrame
+):
+    x = chunk_metadata["x_index"]
+    y = chunk_metadata["y_index"]
+
+    neighbors = {(x - 1, y), (x + 1, y), (x, y - 1), (x, y + 1)}
+    queue = queue.Queue()
+    for neighbor in neighbors:
+        queue.put(neighbor)
+
+    while not queue.empty():
+        neighbor_x, neighbor_y = queue.get()
+        query = (
+            chunks_ref.where("x_index", "==", neighbor_x)
+            .where("y_index", "==", neighbor_y)
+            .limit(1)
+        )
+        chunk_doc = query.get()
+        if not chunk_doc.exists:
+            # Some chunks may not have neighbors in the study area.
+            continue
+
+        neighbor_chunk_metadata = chunk_doc.to_dict()
+        if (
+            neighbor_chunk_metadata.get("id") is None
+            or neighbor_chunk_metadata.get("row_count") is None
+            or neighbor_chunk_metadata.get("col_count") is None
+            or neighbor_chunk_metadata.get("x_ll_corner") is None
+            or neighbor_chunk_metadata.get("y_ll_corner") is None
+            or neighbor_chunk_metadata["x_index"] is None
+            or neighbor_chunk_metadata["y_index"] is None
+        ):
+            raise ValueError(
+                f'Neighbor chunk "{neighbor_chunk_metadata["id"]}" is missing one or more required '
+                "fields: id, row_count, col_count, x_ll_corner, y_ll_corner, x_index, y_index"
+            )
+
+        neighbor_chunk_boundary = _get_chunk_boundary(
+            study_area_metadata, neighbor_chunk_metadata
+        )
+        intersects = False
+        for row in boundary_h3_cells.iterrows():
+            if row["h3_boundary"].intersects(neighbor_chunk_boundary):
+                intersects = True
+                break  # Once an overlap is found, no need to check further
+
+        if intersects:
+            queue.put((neighbor_x - 1, neighbor_y))
+            queue.put((neighbor_x + 1, neighbor_y))
+            queue.put((neighbor_x, neighbor_y - 1))
+            queue.put((neighbor_x, neighbor_y + 1))
```

Comment:
I think we could be a bit more targeted in which neighbors we retrieve. I think we'd only ever need the eight neighbors surrounding the original file as long as the chunks are larger than h3 cells (which they should be.) It looks like this would have us inspecting neighbors-of-neighbors like the chunk two to the right which we shouldn't need. 

We could also look at the boundaries to figure out specifically what we need (e.g. we could see we only need the neighbor to the right and nothing else.) I think we can call that a premature optimization if it's hard, though.

Reply:
IDT either of these optimizations are difficult but would add more complexity to the code. One simpler optimization could be to only read predictions from GCS if the neighboring chunk actually has intersecting H3 cells with the current chunk (rather than reading predictions first and then checking if there are any intersecting H3 cells).

Do you think processing a neighboring chunk adds too much overhead that we should be very targeted in which neighbors we retrieve?

Reply:
How fast is checking the boundaries?
I'd generally suggest correctness & completeness first and worry about performance later.

I think just getting the eight surrounding neighbors (eight because we need to get the neighbor up & to the right, up & to the left, etc along diagonals) should save us grabbing some neighbors we know we don't need (e.g. the neighbor of the neighbor two chunks to the right) but still be simple to implement?

Reply:
WDYM by "How fast is checking the boundaries?" How fast is the processing or how fast would it be to implement?

Regarding the 8 surrounding neighbors- we could do this but we'd be making an assumption about the H3 cell size. I could see a scenario where we start with resolution 13 and then increase the cell size later on. I would prefer dynamically retrieving the neighbors so there aren't any gotchas with changing the resolution size.

Reply:
I was asking how fast does the processing you've implemented run.

I don't think the current logic is quite right. If we always look at the neighbor up, left, down & right, that means we could hop to one neighbor, then hop back to the original chunk (as in, check the neighbor to the right, then hop to its neighbor to the left, which is the original chunk.) So at the least we'd need to keep track of which chunks we've visited to avoid going in circles. 

We also are skipping the neighbors on diagonals, (e.g. up one & left one) which could contain seams. We'd get to them on a more indirect path by moving, say, up & then left but end up visiting more neighbors than we need to.

Ideally we'd use the boundaries to figure out which neighbors to visit. If it's simple to implement, we could figure out based on the boundaries specifically which neighbor we need to visit (e.g. just the one to the right or just the one above.) That's why I was asking how expensive this is to run. We know we'll be processing neighbors we don't need to process. How big of a deal is that? Like, is it order of seconds or minutes? How hard is picking out specific neighbors based on the boundaries to implement? I'd again focus on correctness & completeness first, but if this is easy to do and the boundary checking you've currently implemented is expensive, we might need to think about this now.

One pretty easy to implement improvement would only be looking at part of the neighbor. As in, if the h3 cell is, say, 30 pixels wide, you only need to look at the leftmost 15 pixels (half of 30) of your right neighbor rather than the whole region. But again would focus on performance & optimization later on unless we already know this is expensive.

Reply:
Those are all valid points...

I updated the implementation to only look at the 8 neighboring cells for now. Once we deploy the function and see how fast processing is, we can focus on optimizations.